> building for cloudtests finished: FAILURE https://lcgapp-services.cern.ch/cvmfs-jenkins/job/CvmfsFullBuildDocker/3114/

Something seems to be broken in the Mac build environment:
```
  CC       libcares_la-ares__close_sockets.lo
In file included from ares__close_sockets.c:17:
In file included from ./ares_setup.h:214:
In file included from ./setup_once.h:33:
In file included from /usr/local/include/stdio.h:64:
/usr/local/include/_stdio.h:93:16: warning: pointer is missing a nullability type specifier (_Nonnull, _Nullable, or _Null_unspecified) [-Wnullability-completeness]
        unsigned char   *_base;
                        ^
/usr/local/include/_stdio.h:93:16: note: insert '_Nullable' if the pointer may be null
        unsigned char   *_base;
                        ^
                          _Nullable 
/usr/local/include/_stdio.h:93:16: note: insert '_Nonnull' if the pointer should never be null
        unsigned char   *_base;
                        ^
                          _Nonnull 
/usr/local/include/_stdio.h:138:32: warning: pointer is missing a nullability type specifier (_Nonnull, _Nullable, or _Null_unspecified) [-Wnullability-completeness]
```
with thousands of similar errors all seemingly related to `-Wnullability-completeness`

--- a/cvmfs/download.cc
+++ b/cvmfs/download.cc
@@ -46,6 +46,7 @@
 #include <cstdio>
 #include <cstdlib>
 #include <cstring>
+#include <map>
 #include <set>
 
 #include "atomic.h"
@@ -1537,6 +1538,7 @@ DownloadManager::DownloadManager() {
   opt_proxy_groups_current_ = 0;
   opt_proxy_groups_current_burned_ = 0;
   opt_num_proxies_ = 0;
+  opt_proxy_shard_ = false;
   opt_max_retries_ = 0;
   opt_backoff_init_ms_ = 0;
   opt_backoff_max_ms_ = 0;
@@ -1612,6 +1614,7 @@ void DownloadManager::Init(const unsigned max_pool_handles,
   opt_proxy_groups_current_ = 0;
   opt_proxy_groups_current_burned_ = 0;
   opt_num_proxies_ = 0;
+  opt_proxy_shard_ = false;
   opt_host_chain_current_ = 0;
   opt_ip_preference_ = dns::kIpPreferSystem;
 
@@ -1678,6 +1681,7 @@ void DownloadManager::Fini() {
 
   delete opt_host_chain_;
   delete opt_host_chain_rtt_;
+  opt_proxy_map_.clear();
   delete opt_proxy_groups_;
   opt_host_chain_ = NULL;
   opt_host_chain_rtt_ = NULL;
@@ -2289,6 +2293,7 @@ bool DownloadManager::ProbeGeo() {
     }
   }
 
+  opt_proxy_map_.clear();
   delete opt_proxy_groups_;
   opt_proxy_groups_ = proxy_groups;
   // In pathological cases, opt_proxy_groups_current_ can be larger now when
@@ -2436,6 +2441,7 @@ void DownloadManager::SetProxyChain(
   // From this point on, use set_proxy_list and set_fallback_proxy_list as
   // effective proxy lists!
 
+  opt_proxy_map_.clear();
   delete opt_proxy_groups_;
   if ((set_proxy_list == "") && (set_proxy_fallback_list == "")) {
     opt_proxy_groups_ = NULL;
@@ -2588,8 +2594,11 @@ DownloadManager::ChooseProxyUnlocked(const shash::Any *hash) {
   if (!opt_proxy_groups_)
     return NULL;
 
-  vector<ProxyInfo> *group = current_proxy_group();
-  return &((*group)[0]);
+  uint32_t key = (hash ? hash->Partial32() : 0);
+  map<uint32_t, ProxyInfo *>::iterator it = opt_proxy_map_.lower_bound(key);
+  ProxyInfo *proxy = it->second;
+
+  return proxy;
 }
 
 /**
@@ -2599,13 +2608,45 @@ void DownloadManager::UpdateProxiesUnlocked(const string &reason) {
   if (!opt_proxy_groups_)
     return;
 
+  // Identify number of non-burned proxies within the current group
   vector<ProxyInfo> *group = current_proxy_group();
-  string old_proxy = (*group)[0].url;
-  unsigned select =
-    prng_.Next(group->size() - opt_proxy_groups_current_burned_);
-  swap((*group)[select], (*group)[0]);
-  string new_proxy = (*group)[0].url;
+  unsigned num_alive = (group->size() - opt_proxy_groups_current_burned_);
+  string old_proxy = JoinStrings(opt_proxy_urls_, "|");
 
+  // Rebuild proxy map and URL list
+  opt_proxy_map_.clear();
+  opt_proxy_urls_.clear();
+  const uint32_t max_key = 0xffffffffUL;
+  if (opt_proxy_shard_) {
+    // Build a consistent map with multiple entries for each proxy
+    for (unsigned i = 0; i < num_alive; ++i) {
+      ProxyInfo *proxy = &(*group)[i];
+      shash::Any proxy_hash(shash::kSha1);
+      HashString(proxy->url, &proxy_hash);
+      Prng prng;
+      prng.InitSeed(proxy_hash.Partial32());
+      for (unsigned j = 0; j < kProxyMapScale; ++j) {
+        const std::pair<uint32_t, ProxyInfo *> entry(prng.Next(max_key), proxy);
+        opt_proxy_map_.insert(entry);
+      }
+      opt_proxy_urls_.push_back(proxy->url);
+    }
+    // Ensure lower_bound() finds a value for all keys
+    ProxyInfo *first_proxy = opt_proxy_map_.begin()->second;
+    const std::pair<uint32_t, ProxyInfo *> last_entry(max_key, first_proxy);
+    opt_proxy_map_.insert(last_entry);
+  } else {
+    // Build a map with a single entry for one randomly selected proxy
+    unsigned select = prng_.Next(num_alive);
+    ProxyInfo *proxy = &(*group)[select];
+    const std::pair<uint32_t, ProxyInfo *> entry(max_key, proxy);
+    opt_proxy_map_.insert(entry);
+    opt_proxy_urls_.push_back(proxy->url);
+  }
+  sort(opt_proxy_urls_.begin(), opt_proxy_urls_.end());
+
+  // Report any change in proxy usage
+  string new_proxy = JoinStrings(opt_proxy_urls_, "|");
   if (new_proxy != old_proxy) {
     LogCvmfs(kLogDownload, kLogDebug | kLogSyslogWarn,
              "switching proxy from %s to %s (%s)",
@@ -2757,6 +2798,7 @@ void DownloadManager::CloneProxyConfig(DownloadManager *clone) {
   clone->opt_proxy_groups_current_burned_ = opt_proxy_groups_current_burned_;
   clone->opt_proxy_groups_fallback_ = opt_proxy_groups_fallback_;
   clone->opt_num_proxies_ = opt_num_proxies_;
+  clone->opt_proxy_shard_ = opt_proxy_shard_;
   clone->opt_proxy_list_ = opt_proxy_list_;
   clone->opt_proxy_fallback_list_ = opt_proxy_fallback_list_;
   if (opt_proxy_groups_ == NULL)

--- a/cvmfs/download.cc
+++ b/cvmfs/download.cc
@@ -970,9 +970,10 @@ void DownloadManager::SetUrlOptions(JobInfo *info) {
     // parameters directly
     std::string purl = proxy->url;
     dns::Host phost = proxy->host;
-    ValidateProxyIpsUnlocked(purl, phost);
+    const bool changed = ValidateProxyIpsUnlocked(purl, phost);
     // Current proxy may have changed
-    proxy = current_proxy();
+    if (changed)
+      proxy = current_proxy();
     info->proxy = proxy->url;
     if (proxy->host.status() == dns::kFailOk) {
       curl_easy_setopt(info->curl_handle, CURLOPT_PROXY, info->proxy.c_str());
@@ -1073,13 +1074,15 @@ void DownloadManager::SetUrlOptions(JobInfo *info) {
  * changed, gather new set of resolved IPs and, if different, exchange them in
  * the load-balance group on the fly.  In the latter case, also rebalance the
  * proxies.  The options mutex needs to be open.
+ *
+ * Returns true if proxies may have changed.
  */
-void DownloadManager::ValidateProxyIpsUnlocked(
+bool DownloadManager::ValidateProxyIpsUnlocked(
   const string &url,
   const dns::Host &host)
 {
   if (!host.IsExpired())
-    return;
+    return false;
   LogCvmfs(kLogDownload, kLogDebug, "validate DNS entry for %s",
            host.name().c_str());
 
@@ -1103,7 +1106,7 @@ void DownloadManager::ValidateProxyIpsUnlocked(
       if ((*opt_proxy_groups_)[group_idx][i].host.id() == host.id())
         (*opt_proxy_groups_)[group_idx][i].host = new_host;
     }
-    return;
+    return false;
   }
 
   assert(new_host.status() == dns::kFailOk);
@@ -1131,6 +1134,7 @@ void DownloadManager::ValidateProxyIpsUnlocked(
   opt_num_proxies_ += new_infos.size();
 
   RebalanceProxiesUnlocked();
+  return true;
 }
 
 

--- a/cvmfs/download.h
+++ b/cvmfs/download.h
@@ -473,7 +473,7 @@ class DownloadManager {  // NOLINT(clang-analyzer-optin.performance.Padding)
   void ReleaseCredential(JobInfo *info);
   void InitializeRequest(JobInfo *info, CURL *handle);
   void SetUrlOptions(JobInfo *info);
-  void ValidateProxyIpsUnlocked(const std::string &url, const dns::Host &host);
+  bool ValidateProxyIpsUnlocked(const std::string &url, const dns::Host &host);
   void UpdateStatistics(CURL *handle);
   bool CanRetry(const JobInfo *info);
   void Backoff(JobInfo *info);

--- a/cvmfs/download.h
+++ b/cvmfs/download.h
@@ -11,6 +11,7 @@
 #include <unistd.h>
 
 #include <cstdio>
+#include <map>
 #include <set>
 #include <string>
 #include <vector>
@@ -390,6 +391,7 @@ class DownloadManager {  // NOLINT(clang-analyzer-optin.performance.Padding)
 
   static const unsigned kDnsDefaultRetries = 1;
   static const unsigned kDnsDefaultTimeoutMs = 3000;
+  static const unsigned kProxyMapScale = 16;
 
   DownloadManager();
   ~DownloadManager();
@@ -559,6 +561,18 @@ class DownloadManager {  // NOLINT(clang-analyzer-optin.performance.Padding)
    * The original proxy fallback list provided to SetProxyChain.
    */
   std::string opt_proxy_fallback_list_;
+  /**
+   * Load-balancing map of currently active proxies
+   */
+  std::map<uint32_t, ProxyInfo *> opt_proxy_map_;
+  /**
+   * Sorted list of currently active proxy URLs (for log messages)
+   */
+  std::vector<std::string> opt_proxy_urls_;
+  /**
+   * Shard requests across multiple proxies via consistent hashing
+   */
+  bool opt_proxy_shard_;
 
   /**
    * Used to resolve proxy addresses (host addresses are resolved by the proxy).

--- a/cvmfs/hash.h
+++ b/cvmfs/hash.h
@@ -12,6 +12,7 @@
 #ifndef CVMFS_HASH_H_
 #define CVMFS_HASH_H_
 
+#include <arpa/inet.h>
 #include <stdint.h>
 
 #include <cassert>
@@ -384,6 +385,14 @@ struct Digest {
         return false;
     }
     return true;
+  }
+
+  /**
+   * Get a partial digest for use when only 32 bits are required
+   */
+  uint32_t Partial32() const {
+    const uint32_t *partial = (const uint32_t *)digest;
+    return ntohl(*partial);
   }
 
 

